src/commitment.rs

```diff
@@ -1,58 +1,81 @@
-use std::{io::Read, iter};
+use std::{io::Read, iter, ops};
 
 use digest::{ExtendableOutput, Update};
 use group::Curve;
 use halo2_proofs::arithmetic::{best_multiexp, CurveAffine, CurveExt};
+use log::*;
 use rayon::prelude::*;
-use serde::Serialize;
+use serde::{Deserialize, Serialize};
 use sha3::Shake256;
 
 use crate::util::parallelize;
 
-#[derive(Clone, Debug, Serialize)]
-#[serde(bound(serialize = "C: Serialize"))]
+#[derive(Debug, PartialEq, Eq, thiserror::Error)]
+pub enum Error {
+    #[error("Can't commit too long input: input len: {input_len}, but limit is {limit}")]
+    TooLongInput { input_len: usize, limit: usize },
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
 pub struct CommitmentKey<C: CurveAffine> {
-    ck: Vec<C>,
+    ck: Box<[C]>,
+}
+
+impl<C: CurveAffine> ops::Deref for CommitmentKey<C> {
+    type Target = [C];
+
+    fn deref(&self) -> &Self::Target {
+        &self.ck
+    }
 }
 
 impl<C: CurveAffine> CommitmentKey<C> {
     pub fn default_value() -> C {
         C::identity()
     }
 
+    pub fn len(&self) -> usize {
+        self.ck.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.ck.is_empty()
+    }
+
     pub fn setup(k: usize, label: &'static [u8]) -> Self {
         // This is usually a limitation on the curve, but we also want 32-bit
         // architectures to be supported.
         assert!(k < 32);
         let n: usize = 1 << k;
 
-        let mut shake = Shake256::default();
-        shake.update(label);
-        let mut reader = shake.finalize_xof();
+        let mut reader = Shake256::default().chain(label).finalize_xof();
 
-        let ck_proj: Vec<_> = iter::repeat_with(|| {
-            let mut uniform_bytes = [0u8; 32];
-            reader.read_exact(&mut uniform_bytes).unwrap();
-            uniform_bytes
+        let ck_proj: Box<[_]> = iter::repeat_with(|| {
+            let mut buffer = [0u8; 32];
+            reader.read_exact(&mut buffer).unwrap();
+            buffer
         })
         .take(n)
         .par_bridge()
         .map(|uniform_byte| (C::CurveExt::hash_to_curve("from_uniform_bytes"))(&uniform_byte))
         .collect();
 
-        let mut ck: Vec<C> = vec![C::identity(); n];
+        let mut ck: Box<[C]> = iter::repeat(C::identity()).take(n).collect();
```

Comment:
After we use Box pointer, do you have roughly estimate of the performance gain here?

Reply:
This transformation accomplishes two things.

1. We signal for developers that the given type does not change in runtime after creation. 
2. It definitely does not allocate any extra bytes. For `Vec` capacity can be higher than the actual size, and `Box<[_]>` always stores only how much is needed.

I.e. it's not an array yet, but it's a bit better.
And by the way, we can switch to an array, because we have compile-time constants for size everywhere, but this is out of scope of this task